File: agent/middleware/verify_pr.py
"""After-agent middleware that automatically verifies PRs after creation."""

from __future__ import annotations

import asyncio
import json as _json
import logging
import os
from typing import Any

from langchain.agents.middleware import AgentState, after_agent
from langgraph.config import get_config
from langgraph.runtime import Runtime

from ..tools.verify_pr import verify_pr

logger = logging.getLogger(__name__)

# Environment variable to enable auto-verification
AUTO_VERIFY_ENV_VAR = "OPEN_SWE_AUTO_VERIFY_PR"
DEFAULT_VERIFY_TIMEOUT = int(os.environ.get("PR_VERIFY_TIMEOUT", "600"))


def _extract_pr_number_from_messages(messages: list) -> int | None:
    """Extract PR number from commit_and_open_pr tool result."""
    for msg in reversed(messages):
        if isinstance(msg, dict):
            content = msg.get("content", "")
            name = msg.get("name", "")
        else:
            content = getattr(msg, "content", "")
            name = getattr(msg, "name", "")

        if name == "commit_and_open_pr" and content:
            try:
                parsed = _json.loads(content) if isinstance(content, str) else content
                if isinstance(parsed, dict) and parsed.get("success"):
                    # Extract PR number from pr_url if available
                    pr_url = parsed.get("pr_url", "")
                    # URL format: https://github.com/owner/repo/pull/123
                    if "/pull/" in pr_url:
                        try:
                            return int(pr_url.split("/pull/")[-1].split("/")[0])
                        except ValueError:
                            pass
            except (ValueError, TypeError):
                pass
    return None


async def _verify_pr_after_agent_impl(
    state: AgentState,
    runtime: Runtime,
) -> dict[str, Any] | None:
    """Implementation of the after-agent middleware for PR verification.

    Only runs if:
    1. OPEN_SWE_AUTO_VERIFY_PR environment variable is set to "true"
    2. The agent successfully created a PR via commit_and_open_pr
    3. A PR number can be extracted from the tool result
    """
    # Check if auto-verification is enabled
    if os.environ.get(AUTO_VERIFY_ENV_VAR, "").lower() not in ("1", "true", "yes"):
        return None

    logger.info("Auto-verification middleware started")

    try:
        config = get_config()
        configurable = config.get("configurable", {})
        thread_id = configurable.get("thread_id")

        if not thread_id:
            logger.debug("No thread_id found, skipping auto-verification")
            return None

        messages = state.get("messages", [])
        pr_number = _extract_pr_number_from_messages(messages)

        if not pr_number:
            logger.info("No PR creation detected, skipping auto-verification")
            return None

        logger.info("Auto-verifying PR #%s for thread %s", pr_number, thread_id)

        # Run verification
        result = await asyncio.to_thread(
            verify_pr,
            pr_number=pr_number,
            timeout=DEFAULT_VERIFY_TIMEOUT,
            add_labels=True,
        )

        if result.get("success"):
            logger.info("Auto-verification passed for PR #%s", pr_number)
        else:
            error = result.get("error")
            if error:
                logger.error("Auto-verification failed for PR #%s: %s", pr_number, error)
            else:
                logger.info("Auto-verification found issues in PR #%s", pr_number)

        return {"auto_verification": result}

    except Exception:
        logger.exception("Error in verify_pr_after_agent middleware")

    return None


# Create the decorated middleware function
verify_pr_after_agent = after_agent(_verify_pr_after_agent_impl)

File: src/middleware/verifyPr.ts
/**
 * After-agent middleware that automatically verifies PRs after creation.
 * Mirrors agent/middleware/verify_pr.py
 *
 * Only runs when OPEN_SWE_AUTO_VERIFY_PR is enabled ("true" / "1" / "yes").
 * Scans recent messages for a successful commit_and_open_pr result, extracts
 * the PR number, and invokes verifyPr() in an isolated sandbox.
 */

import type { AIMessage, ToolMessage } from "@langchain/core/messages";
import { createMiddleware } from "langchain";
import { verifyPr } from "../tools/verifyPr.js";

const AUTO_VERIFY_ENV_VAR = "OPEN_SWE_AUTO_VERIFY_PR";
const DEFAULT_VERIFY_TIMEOUT = Number.parseInt(process.env.PR_VERIFY_TIMEOUT ?? "600", 10);

function extractPrNumberFromMessages(messages: (AIMessage | ToolMessage)[]): number | null {
  for (let i = messages.length - 1; i >= 0; i--) {
    const msg = messages[i];
    const name = "name" in msg ? (msg as ToolMessage).name : undefined;
    if (name !== "commit_and_open_pr") continue;

    const content = typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content);
    try {
      const parsed = JSON.parse(content) as Record<string, unknown>;
      if (!parsed.success) continue;

      const prUrl = parsed.pr_url as string | undefined;
      if (prUrl?.includes("/pull/")) {
        const num = Number.parseInt(prUrl.split("/pull/").pop()?.split("/")[0] ?? "", 10);
        if (!Number.isNaN(num)) return num;
      }
    } catch {
      // ignore
    }
  }
  return null;
}

export const verifyPrAfterAgentMiddleware = createMiddleware({
  name: "VerifyPrAfterAgent",
  afterAgent: async (state) => {
    const enabled = (process.env[AUTO_VERIFY_ENV_VAR] ?? "").toLowerCase();
    if (!["1", "true", "yes"].includes(enabled)) return {};

    const configurable = (state as Record<string, unknown>).configurable as
      | Record<string, unknown>
      | undefined;
    const repo = configurable?.repo as { owner: string; name: string } | undefined;

    if (!repo) return {};

    const messages =
      ((state as Record<string, unknown>).messages as (AIMessage | ToolMessage)[]) ?? [];

    const prNumber = extractPrNumberFromMessages(messages);
    if (!prNumber) return {};

    console.log(`[VerifyPrAfterAgent] Auto-verifying PR #${prNumber}`);

    try {
      const result = await verifyPr(prNumber, {
        timeout: DEFAULT_VERIFY_TIMEOUT,
        addLabels: true,
        repoConfig: repo,
      });

      if (result.success) {
        console.log(`[VerifyPrAfterAgent] PR #${prNumber} verification passed`);
      } else {
        console.log(
          `[VerifyPrAfterAgent] PR #${prNumber} verification ${result.status}: ${result.error}`,
        );
      }

      return {};
    } catch (err) {
      console.error("[VerifyPrAfterAgent] Error:", err);
      return {};
    }
  },
});
